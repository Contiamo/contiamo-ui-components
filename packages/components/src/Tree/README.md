The tree component renders a tree structure with collapsable nodes in a filetree-like design.

### Usage

```jsx
<Tree
  trees={[
    {
      label: "Store",
      childNodes: [
        {
          label: "Region",
          childNodes: [
            {
              label: "City",
              tag: "D",
              childNodes: [],
            },
            {
              label: "County",
              tag: "D",
              childNodes: [],
            },
          ],
        },
      ],
    },
    {
      label: "Legal Entity",
      childNodes: [
        {
          label: "LLC",
          tag: "D",
          childNodes: [],
        },
        {
          label: "Inc.",
          tag: "D",
          childNodes: [],
        },
      ],
    },
  ]}
/>
```
