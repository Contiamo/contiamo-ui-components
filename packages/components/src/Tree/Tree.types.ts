/**
 * ITree naming convention is used because it would otherwise clash with the name of the component.
 * (in this case, component cannot be renamed TreeComponent because of styleguidist constraints)
 */
export interface Tree {
  label: string
  tag?: string
  color?: string
  initiallyOpen?: boolean
  disabled?: boolean
  childNodes: Tree[]
  onClick?: () => void
}
