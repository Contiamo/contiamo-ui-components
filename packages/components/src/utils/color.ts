export const isWhite = (color: string) => ["white", "#fff", "#ffffff"].indexOf(color.toLowerCase()) > -1
