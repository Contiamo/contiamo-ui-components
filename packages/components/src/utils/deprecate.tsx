import * as React from "react"
/*
 * Deprecate a component.
 *
 * Usage:
 *
 * const Comp: React.SFC<Props> = props => (
 *   <div />
 * )
 * 
 * export default deprecate<Props>(props => [ 
 *   ...[ "Your component is deprecated" ],
 *   ...(props.oldProp ? [ "oldProp is deprecated, use newProp instead" ] : [])
 * ])(Comp)
 *
 * Typing caveats may arise. For functional components, the easiest fix is
 * to type them explicitly as React.SFC.
 */

function deprecate<P>(createWarning: (props: P) => string[]) {
  return (Component: React.ComponentType<P>): React.ComponentType<P> => {
    return class DeprecatedComponent extends React.Component<P> {
      componentDidMount() {
        const warnings = createWarning(this.props)

        if (process.env.NODE_ENV !== "production" && warnings.length > 0) {
          console.warn(`Operational UI deprecation warnings:\n  * ${warnings.join("\n  * ")}`)
        }
      }

      render() {
        return <Component {...this.props} />
      }
    }
  }
}

export default deprecate
