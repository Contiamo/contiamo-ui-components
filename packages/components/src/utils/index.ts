export { default as deprecate } from "./deprecate"
export * from "./mixins"
export * from "./color"
export { default as withLabel } from "./with-label"
export { default as wrapDefaultTheme } from "./wrap-default-theme" // https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js

export const isModifiedEvent = (event: any) => !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)
