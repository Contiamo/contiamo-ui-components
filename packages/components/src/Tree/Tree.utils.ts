import { Tree } from "./Tree.types"

const pathsEqual = (path1: number[], path2: number[]) => path1.join("-") === path2.join("-")

export const getInitialOpenPaths = (basePath: number[]) => (tree: Tree): number[][] => {
  return [
    ...(tree.initiallyOpen ? [basePath] : []),
    ...tree.childNodes.map((tree, index) =>
      getInitialOpenPaths([...basePath, index])(tree).reduce((paths, accumulator) => [...paths, ...accumulator], []),
    ),
  ]
}

/**
 * Adds or removes a path from a list of paths.
 * togglePath([1, 2])([[1, 2, 3], [0]]) -> [[1, 2, 3], [0], [1, 2]]
 * togglePath([1, 2])([[1, 2], [0]]) -> [[0]]
 */
export const togglePath = (path: number[]) => (paths: number[][]): number[][] => {
  if (paths.length === 0) {
    return [path]
  }
  const [head, ...tail] = paths
  if (pathsEqual(head, path)) {
    return tail
  }
  return [head, ...togglePath(path)(tail)]
}

export const containsPath = (path: number[]) => (paths: number[][]): boolean => {
  if (paths.length === 0) {
    return false
  }
  const [head, ...tail] = paths
  if (pathsEqual(head, path)) {
    return true
  }
  return containsPath(path)(tail)
}
