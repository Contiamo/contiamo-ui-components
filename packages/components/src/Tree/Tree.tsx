import * as React from "react"
import styled from "react-emotion"

import constants, { OperationalStyleConstants, expandColor } from "../utils/constants"
import Icon from "../Icon/Icon"
import { togglePath, containsPath, getInitialOpenPaths } from "./Tree.utils"
import { Tree as ITree } from "./Tree.types"
import SmallNameTag from "../Internals/SmallNameTag"

export interface Props {
  /** An array of tree structures */
  trees: ITree[]
}

export interface State {
  /** Stores all open paths. E.g. if this.state.openPaths is [ [ 0 ], [ 1, 2 ] ], then the first root, and the third child of the second root are open.  */
  openPaths: number[][]
}

const Container = styled("div")`
  display: block;
`

const TreeContainer = styled("div")`
  margin: 2px 0;
`

const TreeChildren = styled("div")`
  margin-left: 28px;
`

const TreeItem = styled("div")`
  display: flex;
  align-items: center;
  padding: 4px 8px;
  width: fit-content;
  min-width: 160px;
  user-select: none;
  ${({
    theme,
    hasChildren,
    hasTag,
    isTopLevel,
    isDisabled,
  }: {
    theme?: OperationalStyleConstants
    hasChildren: boolean
    hasTag: boolean
    isTopLevel: boolean
    isDisabled: boolean
  }) => `
    font-size: ${hasTag ? theme.font.size.fineprint : theme.font.size.small}px;
    font-weight: ${hasTag || isTopLevel ? theme.font.weight.bold : theme.font.weight.regular};
    font-family: ${hasTag ? theme.font.family.code : theme.font.family.main};
    color: ${theme.color.text.dark};
    ${
      isDisabled
        ? "opacity: 0.4;"
        : `
      cursor: pointer;
      :hover {
        background-color: ${theme.color.background.lighter};
      }
    `
    }
    & svg {
      visibility: ${hasChildren ? "visible" : "hidden"};
      color: ${theme.color.text.lightest};
    }
  `};
`

const TreeRecursive: React.SFC<{
  tree: ITree
  path: number[]
  togglePath: (path: number[]) => void
  openPaths: number[][]
}> = ({ tree, path, togglePath, openPaths }) => {
  const isOpen = containsPath(path)(openPaths)
  const { label, tag, disabled, initiallyOpen, childNodes, color, ...treeHtmlProps } = tree
  const tagColor = expandColor(constants, color)
  return (
    <TreeContainer>
      <TreeItem
        {...treeHtmlProps}
        hasTag={Boolean(tree.tag)}
        hasChildren={tree.childNodes.length > 0}
        isTopLevel={path.length < 2}
        isDisabled={Boolean(tree.disabled)}
        onClick={() => {
          treeHtmlProps.onClick && treeHtmlProps.onClick()
          togglePath(path)
        }}
      >
        <Icon left name={isOpen ? "ChevronDown" : "Add"} size={12} />
        {tree.tag && (
          <SmallNameTag color={tagColor} left>
            {tree.tag}
          </SmallNameTag>
        )}
        {tree.label}
      </TreeItem>
      {isOpen &&
        !tree.disabled && (
          <TreeChildren>
            {tree.childNodes.map((tree, index) => (
              <TreeRecursive
                key={index}
                tree={tree}
                path={[...path, index]}
                togglePath={togglePath}
                openPaths={openPaths}
              />
            ))}
          </TreeChildren>
        )}
    </TreeContainer>
  )
}

class Tree extends React.Component<Props, State> {
  readonly state: State = {
    openPaths: this.props.trees
      .map((tree, index) => getInitialOpenPaths([index])(tree))
      .reduce((current, accumulator) => [...current, ...accumulator], []),
  }

  togglePath = (path: number[]) => {
    this.setState(prevState => ({
      openPaths: togglePath(path)(prevState.openPaths),
    }))
  }

  render() {
    return (
      <Container {...this.props}>
        {this.props.trees.map((tree, index) => (
          <TreeRecursive
            key={index}
            tree={tree}
            path={[index]}
            togglePath={this.togglePath}
            openPaths={this.state.openPaths}
          />
        ))}
      </Container>
    )
  }
}

export default Tree
