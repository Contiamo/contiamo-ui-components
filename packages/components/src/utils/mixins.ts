import { Theme } from "@operational/theme"
import { lighten } from "@operational/utils"

export const inputFocus = ({ theme }: { theme: Theme }) => ({
  outline: 0,
  border: "1px solid",
  borderColor: theme.colors.info,
  boxShadow: `0 0 0 3px ${lighten(theme.colors.info, 40)}`
})
