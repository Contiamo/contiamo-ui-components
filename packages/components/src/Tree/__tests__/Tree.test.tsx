import * as React from "react"
import { render } from "enzyme"
import { Tree as ThemelessTree } from "../../index"
import wrapDefaultTheme from "../../utils/wrap-default-theme"
import { togglePath } from "../Tree.utils"

const Tree = wrapDefaultTheme(ThemelessTree)

describe("Tree Component", () => {
  it("Should intialize without problems", () => {
    const output = render(<Tree trees={[]} />)
    expect(output).toMatchSnapshot()
  })
  it("Toggle path - add case", () => {
    expect(togglePath([1, 2])([[1, 2, 3], [0]])).toEqual([[1, 2, 3], [0], [1, 2]])
  })
  it("Toggle path - remove case", () => {
    expect(togglePath([1, 2])([[1, 2], [0]])).toEqual([[0]])
  })
})
