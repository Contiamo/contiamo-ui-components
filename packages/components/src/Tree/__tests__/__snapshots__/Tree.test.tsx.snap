// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Component Should intialize without problems 1`] = `
<div
  class="css-ar73w8-Messages"
/>
`;
